unix/main: Use standard pyexec/repl for unix port.

andrewleech opened the pull request:
This improves repl usage consistency across ports. 
Only enabled when MICROPY_USE_READLINE == 1 (default).

In particular I noticed that `aiorepl` didn't work properly on the unix port currently, it would echo the "just entered line" on a newline each time you hit enter, eg.
```
>>> import aiorepl, asyncio;asyncio.run(aiorepl.task())
Starting asyncio REPL...
--> print("hello")
print("hello")
hello
```
I feel there were also a few other inconsistencies with how I'm used to the repl working on the stm port / via mpremote which are alsos hopefully resolved now.

I've got an optional couple of lines to restore/reset stdio_raw mode when running code - however if raw mode is disabled the aiorepl issue above remains. I think for consistency with other ports it's better to keep stdio in raw mode, but not sure if this is likely to have other side effects?

This PR consilidates the repl functionality and fixes the above aiorepl glitch.

andrewleech committed:
unix/main: Use standard pyexec REPL for unix and windows ports.

This improves REPL usage consistency across ports, by utilizing the pyexec
code for the unix REPL.

Only enabled when MICROPY_USE_READLINE == 1 (the default).

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

dpgeorge commented:
Thanks for this, it's a good clean up.

I actually attempted it before, see #6008.  The approach here is much simpler, so maybe it's a good first step to make.

andrewleech commented:
Ah interesting, I did see some of the return handling and kind of assumed it shouldn't matter - if someone is dealing with the repl (live coding) they're unlikely to care about the exit code?

I didn't consider the SystemExit interrupt at all, haven't looked at how that behaves but probably should!

If running a python file directly then certainly the exit code is important but that shouldn't be impacted by this.

andrewleech commented:
Ah, I might need to read through the other PR discussion a bit more... but in the mean time this is a little off:
```
$ ~/micropython/ports/unix/build-standard/micropython
MicroPython v1.22.0-preview.32.g2c1c7d8b30 on 2023-10-26; linux [GCC 10.2.1] version
Type "help()" for more information.
>>> raise ValueError()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError:
>>> raise SystemExit()
>>>
```

jimmo commented:
See also https://github.com/micropython/micropython/pull/12807

andrewleech commented:
I ran into a tricky issue when trying to fix the unit tests; the banner on Unix port has changed a little with this MR (now consistent with other ports) so unit test exp needed to be updated to match.

I initially tried just copying the results/*.out files to replace the original .exp ones, but the tests still failed! There were inconsistent newlines in the files that tripped up the test.
I narrowed the program down to the pty usage:
* micropython was outputting stdout with '\r\n`
* pty was converting `\n` to `\r\n`, creating a few `\r\r\n`
* the run_tests script does convert `\r\n` back to `\n` but I think it was only doing that once before writing failed test *.out files, then converting a second time to do the actual comparison to *.exp. 
Or it was the offer way around... or maybe subprocess was doing some conversions too.
Either way I was getting some `\r\n` and some `\n` in the test out files.
The same issue is discussed here which is where I found reference to how to disable the pty below conversions: https://www.thecodingforums.com/threads/new-line-conversion-with-popen-attached-to-a-pty.961960/post-5100925

stinos commented on ports/windows/micropython.vcxproj:
```diff
@@ -88,6 +88,7 @@
     <ClCompile Include="@(PyCoreSource)" />
     <ClCompile Include="@(PyExtModSource)" />
     <ClCompile Include="$(PyBaseDir)shared\readline\*.c" />
+    <ClCompile Include="$(PyBaseDir)shared\runtime\pyexec.c" />
```
Super minor but can you move this one line down so it's aplhabetically?

andrewleech replied:
Ah whoops, I honestly thought I was alreadyt putting it in order, eyes blurred readline/runtime and completely missed runtime\gchelper_generic below!

stinos commented on tests/run-tests.py:
```diff
@@ -232,6 +233,11 @@ def send_get(what):
                         # instead of: output_mupy = subprocess.check_output(args, stdin=f)
                         master, slave = pty.openpty()
+                        # disable auto newline conversion on the pty terminal.
+                        attrs = termios.tcgetattr(slave)
```
You're probably aware but this doesn't exist on windows. Isn't there a way to just do a `replace("/r/n", "/n")` before comparing results, or something like that?

andrewleech replied:
Oh, I thought `pty` itself wasn't avaialble either so it would be a moot point? 
I initially thought of just adding _another_ newline replace, but thought fixing the problem at the source was cleaner than further fiddling with test results.

andrewleech replied:
Yeah it looks like these repl_* tests are already skipped on windows master builds https://ci.appveyor.com/project/dpgeorge/micropython/builds/48536058/job/4o16bx3lnvsu7nbn
![image](https://github.com/micropython/micropython/assets/3318786/4de4f577-4ffb-43f1-b309-6e5cfd4b02cc)

dpgeorge commented:
Regarding the comment above about `raise SystemExit`: actually this PR changes the behaviour, now if you `raise SystemExit` from the REPL it does nothing.  Currently on master it will exit MicroPython back to the shell.

We need to retain the existing behaviour, so we don't break existing usage, and because that's how CPython works.

Then it's a separate question how to handle this on bare-metal targets.  But for now we shouldn't change the unix behaviour on this.

andrewleech commented:
> Regarding the comment above about `raise SystemExit`: actually this PR changes the behaviour, now if you `raise SystemExit` from the REPL it does nothing. Currently on master it will exit MicroPython back to the shell.
> 
> We need to retain the existing behaviour, so we don't break existing usage, and because that's how CPython works.
> 

Thanks, yes I hadn't realised this behavior had been changed inadvertantly. I've now fixed it in the latest push.
There was an associates variable here `pyexec_system_exit` which I did a global search for - it was never actually being used (anymore) so I've stripped that out as well. 
Any variable passed to SystemExit is now returned for unix port.
```
anl@STEP: ~ $ micropython
MicroPython v1.24.0-preview.102.g03e992d59d.dirty on 2024-07-09; linux [GCC 9.4.0] version
Type "help()" for more information.
>>> raise SystemExit
anl@STEP: ~ $ echo $?
0
anl@STEP: ~ $ micropython
MicroPython v1.24.0-preview.102.g03e992d59d.dirty on 2024-07-09; linux [GCC 9.4.0] version
Type "help()" for more information.
>>> raise SystemExit(-1)
anl@STEP: ~ $ echo $?
255
```
... oh, maybe one last little bug to fix there...

> Then it's a separate question how to handle this on bare-metal targets. But for now we shouldn't change the unix behaviour on this.

Well it looks to me like this should be treated as a soft reset. 
https://github.com/micropython/micropython/blob/358e501e75c028133cff1045d2062ae31206a22c/extmod/modmachine.c#L66

The current push also works this way:
```
anl@STEP: ~/micropython $ mpremote
Connected to MicroPython at /dev/ttyACM0
Use Ctrl-] or Ctrl-x to exit this shell
MicroPython v1.24.0-preview.104.g161a9c3097.dirty on 2024-07-09; WEACTV20_F411 with STM32F411CE
Type "help()" for more information.
>>> raise SystemExit
MPY: sync filesystems
MPY: soft reboot
MicroPython v1.24.0-preview.104.g161a9c3097.dirty on 2024-07-09; WEACTV20_F411 with STM32F411CE
Type "help()" for more information.
>>>
```

dpgeorge commented:
> There was an associates variable here `pyexec_system_exit` which I did a global search for - it was never actually being used (anymore) so I've stripped that out as well.

Right, it would be a nice clean up to get rid of that variable, and also change the behaviour of `SystemExit` on bare-metal ports to match unix/CPython.

Because this is a change in user-facing behaviour, I've split that change out in #15486.

andrewleech commented:
> prints the parse tree out

Thanks, I wasn't aware of this development feature. I also hadn't noticed the amount of overlap between pyexec/parse_compile_execute function and unix/main/execute_from_lexer() but there's probably not much to be really gained by consolidating the two.

I've made the change now and assume it's working:
```
anl@STEP: ~/micropython $ ./ports/unix/build-coverage/micropython -v -v -vMicroPython v1.24.0-preview.187.g95bce29344.dirty on 2024-08-14; linux [GCC 9.4.0] version
Type "help()" for more information.
>>> import os
----------------
[   1] import_name(14) (n=1)
         id(os)
----------------
File <stdin>, code block '<module>' (descriptor: 7f9fea2a9320, bytecode @7f9fea2a9180 11 bytes)
Raw bytecode (code_info_size=3, bytecode_size=8):
 08 02 01 80 51 1b 02 16 02 51 63
arg names:
(N_STATE 2)
(N_EXC_STACK 0)
  bc=0 line=1
00 LOAD_CONST_SMALL_INT 0
...
```

dpgeorge commented:
It looks like there are still a couple of things missing in pyexec.c that the unix port has:
- support the `compile_only` flag
- set `__file__` when `MICROPY_PY___FILE__` is enabled

dpgeorge commented:
> It looks like there are still a couple of things missing in pyexec.c that the unix port has

This is why #6008 is quite a bit bigger than this PR, and why it was still in progress, because of all these details :smile:

dpgeorge commented:
@andrewleech there are still two of my outstanding points above to address, and also @stinos comment about newlines.  Did you want to address those?  Alternatively I can merge #16141 now to get raw REPL support on the unix port, then this PR can go in after that, when it's ready.

andrewleech commented:
I would very much like to get this PR finished to unblock aiorepl use, I'll reprioritize it in my backlog cleanups effort! 

Previously I didn't like the idea of a blanket replace it newlines like @stinos suggested however after battling inconsistent newline matching in some cmdline tests just the other day I'm now far more interested in the simple/broad fix!

pi-anl committed:
shared/runtime/pyexec: Provide support for compile-only mode.

When `MICROPY_PYEXEC_COMPILE_ONLY` is enabled and the global
`mp_compile_only` is True, code is compiled but not executed.

Also add comprehensive tests for compile-only functionality covering both
successful compilation and syntax error detection.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

pi-anl committed:
shared/runtime/pyexec: Set __file__ for file input when enabled.

When `MICROPY_MODULE___FILE__` is enabled and parsing file input, set the
global `__file__` variable to the source filename. This matches the
behavior of the unix port and provides the current filename to the
executing script.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

pi-anl committed:
tests/run-tests.py: Add general newline normalization function.

Add a general normalize_newlines() function that handles newline variations
(\\r\\r\\n, \\r\\n) to \\n while preserving literal \\r characters that are
part of test content.

This provides a robust solution for cross-platform test compatibility,
particularly addressing PTY double-newline issues that can occur with some
terminal implementations.

The function is applied to all test output before comparison, eliminating
platform-specific newline issues.

Includes a unit test to verify the normalization behavior.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

pi-anl committed:
unix: Enable compile-only mode with shared pyexec REPL.

Provides support for command line `-X compile-only` option on unix port.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

andrewleech commented:
@dpgeorge this should be ready for re-review.

This PR does not include the unhandled exception hooks that you had in https://github.com/micropython/micropython/pull/6008 but I feel that should be a separate feature / pr anyway if it's still needed by others?

There's a minor compatibility update to aiorepl in https://github.com/micropython/micropython-lib/pull/1016 which has been tested with this PR and seems to behave as expected. Without that change aiorepl needs crtl-enter to execute / send a line.

dpgeorge commented on shared/runtime/pyexec.c:
```diff
@@ -116,7 +131,9 @@ static int parse_compile_execute(const void *source, mp_parse_input_kind_t input
         start = mp_hal_ticks_ms();
         #endif
-        mp_call_function_0(module_fun);
+        if (!(exec_flags & EXEC_FLAG_COMPILE_ONLY)) {
```
This compile-only option doesn't work correctly at the REPL, eg:
```
$ ./build-standard/micropython -X compile-only
MicroPython v1.26.0-preview.282.gdcbdeb839 on 2025-06-27; linux [GCC 15.1.1] version
Type "help()" for more information.
>>> 1/2
0.5
>>> 
```
That's wrong, it shouldn't print anything as the answer.

I suggest to just keep this simple and add a compile-time option to enable this:
```c
#if MICROPY_PYEXEC_COMPILE_ONLY
if (mp_compile_only)
#endif
{
    mp_call_function_0(module_fun);
}
```
Then the unix port enables this and makes `mp_compile_only` globally available and declared in `mphalport.h`.

That will mean `pyexec_file_with_flags()` isn't needed, and ports that don't use this feature won't take a hit.

andrewleech replied:
Thanks for the suggestion, I've implemented this and pushed.

dpgeorge commented:
@andrewleech can you please check the CI failures?  Not sure what's going on there.

dpgeorge commented:
Note: the code size increase across most of the ports is due to improved support for `MICROPY_PY___FILE__`: the variable `__file__` is now set whenever a script is executed via pyexec.  So now we have:
- `__file__` is already set on import of scripts (eg `import foo`, then `__file__` would be `foo.py`), that doesn't change with this PR
- when executing boot.py/main.py from the filesystem, `__file__` is set to either `"boot.py"` or `"main.py"` respectively (this is not done for frozen boot.py/main.py though)
- doing `mpremote run test.py` now sets `__file__` to `"<stdin>"`
- at the REPL, `__file__` now exists and is the name of the most recently executed file

I think that's a good improvement, makes things more consistent.

dpgeorge commented on shared/runtime/pyexec.c:
```diff
@@ -64,7 +65,7 @@ static bool repl_display_debugging_info = 0;
 // EXEC_FLAG_IS_REPL is used for REPL inputs (flag passed on to mp_compile)
 static int parse_compile_execute(const void *source, mp_parse_input_kind_t input_kind, mp_uint_t exec_flags) {
-    int ret = 0;
+    mp_int_t ret = 0;
```
I think this `ret` variable should stay as `int`, since that's the return type of this function.

dpgeorge commented on ports/unix/mphalport.h:
```diff
@@ -48,6 +49,12 @@ void mp_hal_set_interrupt_char(char c);
 void mp_hal_stdio_mode_orig(void);
 
+// pyexec/repl needs stdio to be in raw mode, but this may be cleared before running code.
+#if MICROPY_REPL_RESET_RAW_BEFORE_EXEC
```
I don't think this is used anymore??

dpgeorge commented on ports/unix/main.c:
```diff
@@ -297,18 +234,25 @@ static int do_repl(void) {
+    if (ret == 1) {
+        return 0; // success
+    } else if (ret == PYEXEC_FORCED_EXIT) {
+        return FORCED_EXIT; // SystemExit (without exit value for now)
```
The value of `SystemExit` must be propagated out of this function, otherwise `sys.exit(1)` doesn't return an error code to the shell.

andrewleech commented:
@dpgeorge I've addressed your comments and added a couple of basic tests for systemexit - the outputs aren't very interesting but basically show a number in systemexit is shown as a failed (crash) test. Zero or no arg is a successful result.

dpgeorge commented on ports/unix/main.c:
```diff
@@ -297,18 +234,28 @@ static int do_repl(void) {
+    }
 }
 
 static int do_str(const char *str) {
```
It would be possible to replace the `execute_from_lexer()` below with `pyexec_vstr()` (just make a temporary vstr on the C stack and populate it with `str` and `strlen(str)`).

Might be worth doing that as part of this PR?

andrewleech replied:
`execute_from_lexer` is also still used for a couple of other things. I've trimmed it down a bit to remove now-unused functionality but would it be better refactored into a couple of new `pyexec_str_single()` and `pyexec_stdin()` functions in pyexec just to consolidate the location of all this sort of code? These functions are only needed on desktop really so would probably want extra #defines around them.

andrewleech replied:
I've done this as well now thanks

Josverl commented:
Is there a specific reason for
> this is not done for frozen boot.py/main.py though

dpgeorge commented:
> I'm assuming the `__FILE__` setting section is in a different pathway to the "run from frozen" code path. This wasn't intentional certainly. It might not be that easy to fix in a generic way though as I presume frozen code will be referenced by python path rather than the typical file path that's used here.

Yes, that's correct.  There would need to be extra code added to set `__FILE__` for execution of frozen boot/main.

andrewleech commented:
> Yes, that's correct. There would need to be extra code added to set __FILE__ for execution of frozen boot/main.

I'm thinking that would be worth doing too, would it make sense to just report these frozen modules as `<module name>.py` ?

dpgeorge commented:
> I'm thinking that would be worth doing too for consistency at runtime, would it make sense to just report these frozen modules as `<module name>.py` ?

There's code in `py/builtinimport.c` that does it already for import of frozen modules.  Following that logic would be best.

pi-anl committed:
unix/main: Replace execute_from_lexer with pyexec in do_file and do_str.

Consolidates file and string execution to use the standard pyexec interface
for consistency with other ports.

Simplify execute_from_lexer for remaining usage: Remove unused LEX_SRC_VSTR
and LEX_SRC_FILENAME cases, keeping only LEX_SRC_STR for REPL and
LEX_SRC_STDIN for stdin execution.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

dpgeorge commented on ports/unix/main.c:
```diff
@@ -297,22 +226,43 @@ static int do_repl(void) {
+        return ret; // SystemExit with exit value in lower 8 bits
+    } else {
+        return 1; // exception
+    }
```
This code should be factored/shared with `do_file()`.

andrewleech replied:
Updated as suggested thanks

dpgeorge replied:
I think you forgot to commit this change?

andrewleech commented on ports/unix/main.c:
```diff
@@ -297,22 +226,43 @@ static int do_repl(void) {
-    return execute_from_lexer(LEX_SRC_STR, str, MP_PARSE_FILE_INPUT, false);
+    vstr_t vstr;
+    vstr_init(&vstr, strlen(str));
+    vstr_add_strn(&vstr, str, strlen(str));
```
Updated as suggested thanks

dpgeorge replied:
I think you forgot to commit this change?

andrewleech commented:
> @andrewleech can you please rebase this on latest master

done

dpgeorge committed:
shared/runtime/pyexec: Call mp_hal_stdio_mode_orig/raw as appropriate.

This ensures that ctrl-C works on the unix port when executing code at the
REPL.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
It looks like my fix for ctrl-C handling has slightly changed the output for using paste mode, probably due to switch in/out of raw mode.

@andrewleech do you think we can just change the tests .exp file to match the new behaviour?  Or try to make it go back to the previous behaviour?

pi-anl committed:
unix: Enable exit code handling for sys.exit().

Enable `MICROPY_PYEXEC_ENABLE_EXIT_CODE_HANDLING` to propagate `sys.exit()`
exit codes properly. Update `convert_pyexec_result()` to handle return
values where pyexec returns the exit code with `PYEXEC_FORCED_EXIT` flag
set for `SystemExit`. Extract the exit code from the lower 8 bits when the
flag is set, otherwise return as-is (0 for success, 1 for exception).

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

pi-anl committed:
shared/runtime/pyexec: Set PYEXEC_FORCED_EXIT flag for SystemExit.

When `MICROPY_PYEXEC_ENABLE_EXIT_CODE_HANDLING` is enabled, `SystemExit`
now sets the `PYEXEC_FORCED_EXIT` flag in addition to the exit code. This
allows the REPL to properly detect and exit when SystemExit is raised,
while still preserving the exit code in the lower bits.

Fixes `repl_lock.py` test which expects REPL to exit on `SystemExit`.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

andrewleech commented:
The unix ctrl-c tests have been split out from this PR again into https://github.com/micropython/micropython/pull/18435

dpgeorge reviewed (approved):
I tested this thoroughly and fixed a minor issue: ctrl-C was not properly handled when executing a string via `micropython -c "import time; time.sleep(10)"`.
